crypto/asn1/tasn_dec.c

```diff
@@ -738,11 +744,29 @@ static int asn1_d2i_ex_primitive(ASN1_VALUE **pval, const unsigned char **in,
       p += plen;
     }
   } else if (cst) {
-    // This parser historically supported BER constructed strings. We no
-    // longer do and will gradually tighten this parser into a DER
-    // parser. BER types should use |CBS_asn1_ber_to_der|.
-    OPENSSL_PUT_ERROR(ASN1, ASN1_R_TYPE_NOT_PRIMITIVE);
-    return 0;
+    if (utype == V_ASN1_NULL || utype == V_ASN1_BOOLEAN
+                || utype == V_ASN1_OBJECT || utype == V_ASN1_INTEGER
+                || utype == V_ASN1_ENUMERATED) {
+      // These types only have primitive encodings.
+      OPENSSL_PUT_ERROR(ASN1, ASN1_R_TYPE_NOT_PRIMITIVE);
+      return 0;
+    }
+
+    // Should really check the internal tags are correct but some things
+    // may get this wrong. The relevant specs say that constructed string
+    // types should be OCTET STRINGs internally irrespective of the type.
+    // So instead just check for UNIVERSAL class and ignore the tag.
+    if (!asn1_collect(&buf, &p, plen, inf, -1, V_ASN1_UNIVERSAL, 0)) {
+      goto err;
+    }
+    len = buf.length;
+    // Append a final null to string.
+    if (!BUF_MEM_grow_clean(&buf, len + 1)) {
+      OPENSSL_PUT_ERROR(ASN1, ERR_R_MALLOC_FAILURE);
```

Comment:
nit: `OPENSSL_malloc` gets called under the hood, so setting the error here is redundant, right?

Reply:
Correct, since the `OPENSSL_malloc` change was made after the commit that removed support for BER constructed strings. Will remove!